### PR TITLE
Account management

### PR DESCRIPTION
**Pull request by SirRichardFrancis**
## About The Pull Request

Added PDA app for managing accounts.

<details>
<summary>
Description
</summary>

Buttons:

For any account:

- "Enter Tax ID" enter account number and password.
- "Log in" actually login with your account.
- "Browse logs" lists all transactions, nothing interesting here. 
- "Transfer funds" allow to transfer money, have to know target account number.
- "Log out" terminates session.

For personal accounts:

- "Resign" breaks link with current employer and removes salary. Basically quitting a job. 

For department accounts:

- "Disavow" breaks link with target employee, removing them from payroll. 
- "Link employee account" pretty much reversed "Disavow", need target account number.
- "Set wage" choose salary for an employee. That will switch "Handled" to "Manually".
- "Reset wage" reset to default / recalculate salary. That will switch "Handled" to "Automatically".

For aster's accounts (captain, first officer, merchant, command and guild): 

- "Register new account" input a name for new account and get login and password. Cost 500 credits.

Stats:

For any account:

- "Balance" - how much money account currently have.
- "Income" - how much money this account paid each hour.
- "Debt" - how much money employer was meant to pay, but didn't.
- "Handled" - mostly used for department accounts that recalculate their salary when crew enters/leaves cryo pods.
- "Employer" - who is to pay salary to this account.

For department accounts:

- "Employee" - linked account's or it owner's name .
- "Wage" - how much they're meant to be paid each hour.
- "Debt" - how much they should have been paid, but haven't been due to insufficient funds. This will be added to next hour's wage.
- "Handled" - "Manually" will stop automatic "Wage" increase/decrease from happening. "Automatically" is default state.

</details>

<details>
<summary>
Screenshots
</summary>

image

image

image

image

image

image

image

</details>

## Why It's Good For The Game

You know why.

## Changelog
:cl:
add: up-to-date taxquickly app, allows for account creation and salary management.
add: bugs
refactor: reworked "payday"
/:cl:

### CODE REVIEW

**Comment by Anonmare** [comment]
I strongly approve, having a means to create bank accounts and manage automatic payments has been a long time coming, especially now thatfactions have a tangible reason to remain in the Captain's good graces in order to keep getting paid.

**Comment by Voldirs** [comment]
Now that's a good shit

**Comment by Anonmare** [comment]
Can registering a new account be set-up so that it can be defined as a personal or a departmental account? Like, say, if I wanted to create my own vagabond merc group with a bank acc that'll pay them?

**Comment by SirRichardFrancis** [comment]
> Can registering a new account be set-up so that it can be defined as a personal or a departmental account? Like, say, if I wanted to create my own vagabond merc group with a bank acc that'll pay them?

It currently can not. There is many functions missing (more popups, taxation, good UI) that I would like to add, but alas, very short on time. Maybe I'll make it before Friday. If not, then in next PR some day.

**Comment by TheShown911** [comment]
Great PR, currently unable to review the code, hope it doesn't slip past on Friday.

**Comment by SirRichardFrancis** [comment]
Can make departments now, for a fee of 2k credits. Feature seems to work, but have no time to test properly.

File: code/modules/modular_computers/file_system/programs/generic/taxes.dm [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+		T.date = current_date_string
+		T.time = stationtime2text()
+		T.source_terminal = account.department_id? "[account.get_name()]" : "Asters Guild Representative [account.get_name()]"
+
+		M.transaction_log.Add(T)
+		all_money_accounts.Add(M)
+		personal_accounts.Add(M)
+
+		charge_to_account(account.account_number, owner_name, "Account registration fee", name, registration_fee)
+		popup_message = "<b>Account created!</b><br> Make sure to copy following information now.<br> You won’t be able to see it again!<br> Tax ID: [M.account_number]<br> Pin code: [M.remote_access_pin]"
```
Comment (hyperioo) [inline code/modules/modular_computers/file_system/programs/generic/taxes.dm:224]: Don't know if this `’` is supported

File: code/modules/modular_computers/file_system/programs/generic/taxes.dm [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+		popup = TRUE
+		return TOPIC_REFRESH
+	
+	return TOPIC_HANDLED
+
+
+/datum/nano_module/program/tax/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = NANOUI_FOCUS, var/datum/topic_state/state = GLOB.default_state)
+	var/list/data = host.initial_data()
+//	var/datum/computer_file/program/tax/PRG = program
+	data["stored_login"] = account_num? account_num : ""
```
Comment (hyperioo) [inline code/modules/modular_computers/file_system/programs/generic/taxes.dm:234]: ` ? ` for clarity, noting it only once but it's the same for following lines

**Comment by Jan-Blasiak** [comment]
### **All hail SirRichardFrancis!**

File: code/controllers/subsystems/economy.dm [hunk 3]
```diff
@@ -21,179 +21,86 @@ SUBSYSTEM_DEF(economy)
...
+			if(R)
+				payroll_mail_account_holder(R, "[ED.name] account", amount_to_pay)
+		else
+			A.debt += A.wage
+			ED.total_debt += A.wage
+			// payroll_mail_where_is_my_money
+
+	// Mail commanding officers and politely ask "Where's the fucking money, shithead?"
+	for(var/i in GLOB.all_departments)
+		var/datum/department/D = GLOB.all_departments[i]
```
[the remark this answers is not in the document]
Comment (SirRichardFrancis) [reply code/controllers/subsystems/economy.dm:93]: No? Have to specify that datum is `/datum/department` or byond will be unable to read variables that only `/datum/department` have.

File: code/controllers/subsystems/economy.dm [hunk 4]
```diff
@@ -21,179 +21,86 @@ SUBSYSTEM_DEF(economy)
...
+				deposit_to_account(A, D.funding_source, "Payroll Funding", "Hansa payroll system", amount_to_pay)
+				paid_external += amount_to_pay
 
+			if(FUNDING_INTERNAL)
+				if(amount_to_pay <= EA.money)
+					transfer_funds(EA, A, "Payroll Funding", "CEV Eris payroll system", amount_to_pay)
+					paid_internal += amount_to_pay
+					ED.total_debt -= A.debt
+					A.debt = 0
+				else
```
[the remark this answers is not in the document]
Comment (SirRichardFrancis) [reply code/controllers/subsystems/economy.dm:60]: What? Rephrase that, I don't understand what you mean.

File: code/modules/economy/Accounts.dm [hunk 5]
```diff
@@ -130,7 +141,16 @@
...
 	all_money_accounts.Add(M)
-
+	personal_accounts.Add(M)
+
+	// Increase personnel budget of our department, if have one
+	if(department && wage)
+		var/datum/money_account/EA = department_accounts[department]
+		var/datum/department/D = GLOB.all_departments[department]
+		if(D && EA)
+			D.budget_personnel += wage
```
[the remark this answers is not in the document]
Comment (SirRichardFrancis) [reply code/modules/economy/Accounts.dm:151]: You can't create an account with a wage, it have to be assigned by account from which that wage will be paid.
For roundstart accounts only "default" wages from `/datum/job` or `/datum/department` being used, and while department head can set employee wage to 99999 credits, that will not let them syphon the money.
`budget_personnel` also can't be modified by players and serve to determine how much money department account should receive in "authomatic handling" mode, it's like payday system that we have now.

File: code/modules/modular_computers/file_system/programs/generic/taxes.dm [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+			if("Department")
+				registration_fee = department_registration_fee
+				is_department = TRUE
+
+			if("Cancel")
+				return TOPIC_HANDLED
+
+		if(account.money < registration_fee)
+			popup_message = "<b>An error has occurred.</b><br> Can't afford account registration fee.<br> Try again when you're a little richer."
```
[the remark this answers is not in the document]
Comment (SirRichardFrancis) [reply code/modules/modular_computers/file_system/programs/generic/taxes.dm:179]: Addressed.

File: code/modules/modular_computers/file_system/programs/generic/taxes.dm (hunk 1 above)
Comment (SirRichardFrancis) [reply code/modules/modular_computers/file_system/programs/generic/taxes.dm:224]: Replaced with `\'`.

File: code/modules/modular_computers/file_system/programs/generic/taxes.dm (hunk 2 above)
Comment (SirRichardFrancis) [reply code/modules/modular_computers/file_system/programs/generic/taxes.dm:234]: Addressed.

**Review by Firefox13 — APPROVED** [review]

**Pull request merged**